### ISSUE TITLE
Implement dismissSoftInput() for search-bar

### ISSUE THREAD

**Issue by emiloberg** [NativeScript/NativeScript#849]
Feature request:

The search-bar has no way of dismissing an open keyboard. This makes the search-bar a quite unusable as the normal user pattern is that the user search for something and then press the item and gets navigated there. On Android (at least on >= 5.x), the open keyboard will continue to stay open, even on the new page. On iOS it will close.

the `editable-text-base`, and therefore the `text-field`, has a `dismissSoftInput`. Can this be implemented on the search-bar as well?

Related, it should probably have the `focus()` method as well. 

Or am I missing a workaround?

**Comment by TobiasHennig** [comment]
I have the same problem and discover a bit of a workaround (iOS only):

**Option 1: Dismiss input with a "Cancel" button**
![image](https://cloud.githubusercontent.com/assets/3656893/10193022/b573b606-6780-11e5-8727-8861f2191d38.png)
```javascript
var searchBar = page.getViewById("search-bar");
if (applicationModule.ios) {
    searchBar.ios.showsCancelButton = true;
}
```

**Option 2: Dismiss input programmatically**
```javascript
var searchBar = page.getViewById("search-bar");
if (applicationModule.ios) {
    searchBar.ios.endEditing(true);
}
```

**Comment by emiloberg** [comment]
Good iOS workaround! Now we just need an Android one as well

**Comment by emiloberg** [comment]
Alright @TobiasHennig, thanks for getting me to look at the android documentation.

This seems to do the job.

```
var searchBar = page.getViewById("search-bar");
if (searchBar.ios) {
    searchBar.ios.endEditing(true);
} else if (searchBar.android) {
    searchBar.android.clearFocus();
}
```

**Comment by emiloberg** [comment]
And for completness.

As the search-bar inherits from `View` you can call `focus()` to set focus on the search-bar and that way make the keyboard visible.
```
var searchBar = page.getViewById("search-bar");
searchBar.focus()
}
```

Footnote: Android seem to set focus on the search-bar when navigating to a page with such on it. iOS does however not.

**Comment by leocaseiro** [comment]
Hi, I'd like to know how to implement dismissSoftInput for searchBar and how it works.
Any example running? 
May I add this to  `.on(searchBarModule.SearchBar.clearEvent)`?

I've tried the code bellow, but nothing happens on IOS device:

```javascript
exports.pageLoaded = function() {
    var searchBarModule = require('ui/search-bar');
    var searchBar = page.getViewById('searchid');
    
    searchBar.on(searchBarModule.SearchBar.clearEvent, function(args) {
        console.log('-- Cleanned Search for Contacts');
        searchBar.dismissSoftInput();
    });
};
```

I also tried:

```javascript
    searchBar.on(searchBarModule.SearchBar.clearEvent, function(args) {
        if (searchBar.ios) {
            console.log('searchBar.ios');
            searchBar.ios.endEditing(true);
        } else if (searchBar.android) {
            searchBar.android.clearFocus();
        }
    });
```

My last tried was similar the implementation on [editable-text-base.ios.ts](https://github.com/NativeScript/NativeScript/blob/b2c182fd7a2b150098731750045955aa86de2aba/ui/editable-text-base/editable-text-base.ios.ts#L7):


```javascript
    searchBar.on(searchBarModule.SearchBar.clearEvent, function(args) {
        if (searchBar.ios) {
            console.log('searchBar.ios');
            searchBar.ios.resignFirstResponder();
        } else if (searchBar.android) {
            searchBar.android.clearFocus();
        }
    });
```

**Comment by EddyVerbruggen** [comment]
I like that cancel button thingy! For anyone using that: you can also make it appear animated, like this: `searchBar.ios.setShowsCancelButtonAnimated(true, true);`

**Comment by blaur** [comment]
Anyone who have a solution for android? Mine still seems to be behaving very strainge no matter the option I choose from in here.

**Comment by dzfweb** [comment]
Still need for a solution in android.

**Comment by dlucidone** [comment]
I'm facing the same problem with search bar. The focus is unset but then on tab view change it gets focus and sometimes the keyboard pops up.Dont know whats the matter with this.

**Comment by grasshoppermouse** [comment]
Same problem on android. I've tried many different solutions that folks have posted here and on stackoverflow. Some crash my app, others don't work.

**Comment by dlucidone** [comment]
Really want to fix this keyboard action

**Comment by dzfweb** [comment]
I solved the keyboard opening, doing the following:

**component.html**
`<SearchBar (loaded)="onSearchBarLoaded($event)"></SearchBar>`

**component.ts**
```
onSearchBarLoaded(event) {
        if (event.object.android) {
            event.object.dismissSoftInput();
            event.object.android.clearFocus();
            event.object.android.setFocusable(false);
        }
    }
```

**Comment by dlucidone** [comment]
@dzfweb I'm having problem in case of tab view and pull to refresh. each time i do any of the thing it gets poped up

**Comment by dzfweb** [comment]
I have the same case of you, i have 3 tabs, each with a SearchBar. 
While navigating on it, the keyboard still not open.

**Comment by dlucidone** [comment]
@dzfweb I'll try and see if it works. Thanks

**Comment by grasshoppermouse** [comment]
@dzfweb, your solution seems to work for me. Thanks!

**Comment by EddyVerbruggen** [comment]
@dzfweb Thanks, that worked for me as well, although I needed 2 tweaks (on Android 4.4.4) I'd like to share:

```js
  onSearchBarLoaded(event) {
    if (event.object.android) {
      setTimeout(() => {
        event.object.dismissSoftInput();
        event.object.android.clearFocus();
      }, 0);
    }
  }
```

1. I had to wrap it in a timeout, but that's probably depending on when exactly the SearchBar is rendered.
2. I had to remove `event.object.android.setFocusable(false);` as that required me to tap the field twice before being able to enter anything in it. No side effects thus far.
